### PR TITLE
Refactor manager class

### PR DESCRIPTION
**Pull request by su-chang**
Pick up `whatsapp` related methods from `Manager` class to `whatsapp` directory which contain these classes:

- WhatsAppBase
  - WhatsAppManager
  - LoginEventHandler
  - GroupEventHandler
  - MessageEventHandler

### CODE REVIEW

**Comment by hcfw007** [comment]
I think manager refactor and event support should be divided into 2 PRs

**Comment by bung87** [comment]
`src/whatsapp/whatsapp-base.ts` and `src/whatsapp/whatsapp-manager.ts` less than 200 lines , but seperate to 2 files.

**Comment by su-chang** [comment]
> `src/whatsapp/whatsapp-base.ts` and `src/whatsapp/whatsapp-manager.ts` less than 200 lines , but seperate to 2 files.

The `WhatsAppBase` will extends event listener `EE`, and other related classes extends it.

File: src/whatsapp/whatsapp-base.ts [hunk 1]
```diff
@@ -0,0 +1,38 @@
+/* eslint-disable no-case-declarations */
+/* eslint-disable import/no-duplicates */
+import { EventEmitter as EE } from 'ee-ts'
+import { WA_ERROR_TYPE } from '../exceptions/error-type.js'
+import WAError from '../exceptions/whatsapp-error.js'
+
+import type { ManagerEvents } from '../manager-event.js'
+import type { Manager } from '../manager.js'
+
+export default class WhatsAppBase extends EE<ManagerEvents> {
```
Comment (bung87) [inline src/whatsapp/whatsapp-base.ts:10]: seems this class does not need extends from `EE<ManagerEvents>`

File: src/whatsapp/message-event-handler.ts [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+  WhatsAppMessage,
+  WhatsAppMessagePayload,
+} from '../schema/whatsapp-type.js'
+import { RequestPool } from '../request/requestPool.js'
+import { withPrefix } from '../logger/index.js'
+import WhatsAppBase from './whatsapp-base.js'
+
+const logger = withPrefix(`${PRE} MessageEventHandler`)
+
+export default class MessageEventHandler extends WhatsAppBase {
```
Comment (bung87) [inline src/whatsapp/message-event-handler.ts:28]: no need extends full ManagerEvents interfaces.

File: src/whatsapp/login-event-handler.ts [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+  ClientSession,
+  BatteryInfo,
+  WAStateType,
+} from '../schema/whatsapp-type.js'
+import { withPrefix } from '../logger/index.js'
+import WhatsAppBase from './whatsapp-base.js'
+
+const logger = withPrefix(`${PRE} BotEventHandler`)
+
+export default class LoginEventHandler extends WhatsAppBase { // FIXME: I have no good idea for this class name.
```
Comment (bung87) [inline src/whatsapp/login-event-handler.ts:33]: no need extends full ManagerEvents interfaces.

File: src/whatsapp/group-event-handler.ts [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+  genRoomAnnounce,
+  genRoomJoinEvent,
+  genRoomTopicEvent,
+} from '../pure-function-helpers/room-event-generator.js'
+import { withPrefix } from '../logger/index.js'
+import WhatsAppBase from './whatsapp-base.js'
+
+const logger = withPrefix(`${PRE} GroupEventHandler`)
+
+export default class GroupEventHandler extends WhatsAppBase {
```
Comment (bung87) [inline src/whatsapp/group-event-handler.ts:24]: no need extends full ManagerEvents interfaces.

File: src/whatsapp/message-event-handler.ts (hunk 2 above)
Comment (su-chang) [reply src/whatsapp/message-event-handler.ts:28]: So you means that we need a new interface `MessageEvents` like `ManagerEvens` here?

File: src/whatsapp/group-event-handler.ts (hunk 4 above)
Comment (su-chang) [reply src/whatsapp/group-event-handler.ts:24]: ditto

File: src/whatsapp/message-event-handler.ts (hunk 2 above)
Comment (bung87) [reply src/whatsapp/message-event-handler.ts:28]: I think this class only need implement part of ManagerEvents interface.

File: src/whatsapp/whatsapp-base.ts (hunk 1 above)
Comment (bung87) [reply src/whatsapp/whatsapp-base.ts:10]: but these classes no need full ManagerEvents interfaces and may not need extends `WhatsAppBase`

File: src/whatsapp/message-event-handler.ts (hunk 2 above)
Comment (su-chang) [reply src/whatsapp/message-event-handler.ts:28]: > I think this class only need implement part of ManagerEvents interface.

OK, do you agree about creating a new interface and named `MessageEvents`?

Comment (bung87) [reply src/whatsapp/message-event-handler.ts:28]: sure, please.

File: src/whatsapp/whatsapp-base.ts (hunk 1 above)
Comment (su-chang) [reply src/whatsapp/whatsapp-base.ts:10]: I have think this question again.

> but these classes no need full ManagerEvents interfaces and some may not need extends `WhatsAppBase`

I agree with it.

But we also need a base class for some attributes, likes that `manager` `getBotId()` .etc

Do you have some good idea about that? If we make these classes with full `ManagerEvent`, and then what will happened or does it matter?

Maybe we should find a good way to resolve these problems. And lets talk about this problem here: https://github.com/wechaty/puppet-whatsapp/pull/258

**Comment by su-chang** [comment]
This PR will be closed now, and it has already divided into 2 PRs: #257 and #258

**Pull request closed without merging**